### ISSUE TITLE
[BUG] onSubmit/validate/validateOnBlur use out of date values after react reconsilation

### ISSUE THREAD

**Issue by ForbesLindesay** [final-form/react-final-form#128]
### What is the current behavior?

If you replace the `onSubmit` function after the initial render, the new function will be ignored because react-final-form takes that configuration on startup.  Here is an example that makes the issue fairly clear.

```js
class App extends React.Component {
  state = { mounted: false };
  componentDidMount() {
    this.setState({ mounted: true });
  }
  render() {
    return (
        <Form
          onSubmit={
            this.state.mounted ? (
              () =>
                window.alert("This is the on submit that should fire")
            ) : (
              () =>
                window.alert("This is the on submit that actually fires")
            )
          }
          render={({ handleSubmit, reset, submitting, pristine, values }) => (
            <form onSubmit={handleSubmit}>
              <button type="submit" disabled={submitting || pristine}>
                Submit
              </button>
            </form>
          )}
        />
    );
  }
}
```

It can be quite a bit more subtle though.  The following has the same issue because the original arrow function has captured an old version of the state:

```js
class App extends React.Component {
  state = { mounted: false };
  componentDidMount() {
    this.setState({ mounted: true });
  }
  render() {
    const mounted = this.state.mounted;
    return (
        <Form
          onSubmit={
            () => (mounted
               ? window.alert("This is the on submit that should fire")
               : window.alert("This is the on submit that actually fires")
            )
          }
          render={({ handleSubmit, reset, submitting, pristine, values }) => (
            <form onSubmit={handleSubmit}>
              <button type="submit" disabled={submitting || pristine}>
                Submit
              </button>
            </form>
          )}
        />
    );
  }
}
```

Another more subtle way that this can occur is if you replace the entire form with another form - if this is done in place, react will think it can update the existing form, rather than creating a new one.

### Sandbox Link

[![Edit 🏁 React Final Form - Simple Example](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/v6z0m2o767)

### Suggested fixes

There are two ways this could be resolved:

1. Add methods to final-form so that the onSubmit/validate/validateOnBlur config options can be updated.  This update can then happen in `componentWillReceiveProps` alongside the existing update to initial state.  This needs to remain a fairly "cheap" operation, since arrow functions would trigger this on every render.
2. "Proxy" the methods by defining methods that look like `_onSubmit = data => this.props.onSubmit(data)` on the `ReactForm` instance and then use those in place of directly referencing `this.props.onSubmit`.  That way you would always get an up-to-date reference.  The issue with this is that final-form does different things depending on the `length` of `onSubmit` and that can't be replicated here.

Although this issue *is* avoidable, it is going to lead to some really subtle bugs that are hard to track down, so I think it should be fixed.

**Comment by erikras** [comment]
Love your final sentence. 👍

**Comment by ForbesLindesay** [comment]
Do you have any thoughts on these two approaches?

**Comment by erikras** [comment]
Published fix in [`v4.3.0`](https://github.com/final-form/final-form/releases/tag/v4.3.0).

**Comment by erikras** [comment]
Well, reopening until #183 lands.

**Comment by erikras** [comment]
Published fix in [`v3.2.1`](https://github.com/final-form/react-final-form/releases/tag/v3.2.1).
